plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/GeneratorHelpers.kt

```diff
@@ -503,8 +503,10 @@ interface IrBuilderExtension {
     fun findEnumValuesMethod(enumClass: ClassDescriptor): IrFunction {
         assert(enumClass.kind == ClassKind.ENUM_CLASS)
         return compilerContext.referenceClass(enumClass.fqNameSafe)?.let {
-            it.owner.functions.find { it.origin == IrDeclarationOrigin.ENUM_CLASS_SPECIAL_MEMBER && it.name == Name.identifier("values") }
-                ?: throw AssertionError("Enum class does not have .values() function")
+            it.owner.functions.find { f ->
+                (f.origin == IrDeclarationOrigin.ENUM_CLASS_SPECIAL_MEMBER || (it.owner.origin == IrDeclarationOrigin.IR_EXTERNAL_JAVA_DECLARATION_STUB && f.origin == IrDeclarationOrigin.IR_EXTERNAL_DECLARATION_STUB))
+                        && f.name == Name.identifier("values")
```

Comment:
I was just "мимо крокодил", but noticed that here you most likely need just to check the function signature without any origins. I.e. check that the name is `"values"` (already done), and also that there's no extension receiver parameter, and `valueParameters.isEmpty()`.

Origins are just a very unstable concept, and moreover, with the current variant you might accidentally pick up some irrelevant function named "values" in a Java enum.

Reply:
Also, every enum is guaranteed to have the values functions, so you might as well change `find` to `single` to catch any future bugs.

Reply:
@shanshin Please just use `single`, as I mentioned above, instead of manually filtering and checking that there's only one element. Also, you need to check that the function has no extension receiver parameter (also mentioned above)

Reply:
@udalov using `single` function forces to replace error message text and I would not like to change this behavior.

Reply:
Then you can use `singleOrNull` + custom error :)

Reply:
Indeed, you need to check that `it.dispatchReceiverParameter == null` (and probably `extensionReceiverParameter` too, just to to be sure)

You can use `singleOrNull { ... } ?: error("custom message")`. There's barely any need to distinguish between no and multiple `values` methods